### PR TITLE
Define new args structure for registering speed bump

### PR DESCRIPTION
**Pull request by goldenapples**
See #36.

Work in progress - needs more abstraction and fleshing out, as well as a lot of tests to support it. My goal is to build a structure where all of the old arguments will still work as intended by mapping to new arguments. In places where that would introduce too much complexity - and maybe elsewhere, seeing as we're the only ones using this so far, it might be OK to break backward compatibility and just ship theme changes along with the submodule update.

### CODE REVIEW

**Comment by danielbachhuber** [comment]
> My goal is to build a structure where all of the old arguments will still work as intended by mapping to new arguments. In places where that would introduce too much complexity - and maybe elsewhere, seeing as we're the only ones using this so far, it might be OK to break backward compatibility and just ship theme changes along with the submodule update.

Let's just make breaking changes. We haven't shipped a release of the open source plugin yet, so I'd be fine changing the arguments.

**Comment by goldenapples** [comment]
> Let's just make breaking changes.

Sounds good to me.

**Comment by goldenapples** [comment]
@noppanit or @danielbachhuber - can you look over this general structure and let me know if you have ideas for improving on it?

This is just a first pass - everything in here needs to be supported by tests, and I fully expect to find a lot of bugs once I start building out the test for it, but I'd like to get a sense of whether this structure seems like it will meet our needs and if it looks extensible and logical enough...

File: speed-bumps.php [hunk 1]
```diff
@@ -132,33 +146,100 @@ public function insert_speed_bumps( $the_content ) {
...
+			// Rules which govern the content as a whole
+			'minimum_content_length' => array(
+				'paras' => 8,
+				'chars' => 1200,
+			),
+
+			// Positional rules: distance from start, end, and other elements
+			'from_start' => array(
+				'paras' => 3,
+				'words' => 75,
```
Comment (danielbachhuber) [inline speed-bumps.php]: Is this an `OR` comparison or `AND` ?

**Comment by danielbachhuber** [comment]
:100: Love it

File: speed-bumps.php (hunk 1 above)
Comment (goldenapples) [inline speed-bumps.php]: I'm treating it as an `AND`. Every one of those conditions has to be met in order to insert a speed bump. This approach should address conditions where the first three paragraphs are short sentences. So a speed bump can't be inserted until at least 75 words in, which could be three paragraphs or 5 or more...

Comment (danielbachhuber) [inline speed-bumps.php]: Cool. Can we use `paragraphs` instead of `paras`, and `characters` instead of `chars`? The former is more precise.

**Comment by goldenapples** [comment]
OK, all tests pass now, and I added coverage for a lot of the new constraints . Fixed a bunch of bugs and inconsistencies in implementation. @noppanit @danielbachhuber - either of you want to review this and give me feedback on it?

**Comment by noppanit** [comment]
The code looks good to me. I think this one is related to this GH https://github.com/fusioneng/speed-bumps/issues/5

File: speed-bumps.php [hunk 2]
```diff
@@ -132,33 +146,100 @@ public function insert_speed_bumps( $the_content ) {
...
+		// XXX: These old API options need to be handled as well.
+			//'minimum_content_length' => 1200,
+			//'paragraph_offset' => 0,
+			//'element_constraints' => array(
+				//'iframe',
+				//'oembed',
+				//'image',
+				//),
+			//'minimum_space_from_other_inserts' => 1,
+			//);
```
Comment (danielbachhuber) [inline speed-bumps.php]: Are we dropping this?

**Comment by danielbachhuber** [comment]
@goldenapples Looks good. Can you:

1. Resolve the merge conflict
2. Merge to master
3. Set up a PR for the theme which cleans up our existing speed bumps?

**Comment by goldenapples** [comment]
Merged master with the latest tests and made them all pass. Merging this now, and will follow with an update in fusion theme to use the new args structure.

**Pull request merged**